## BUSINESS QUESTIONS
1. Berapa rata-rata pengeluaran pengguna per bulan?
2. Pada kuartal mana pengeluaran mencapai jumlah tertinggi dalam setahun?
3. rasio pengeluaran terhadap kategori kebutuhan vs keinginan?
3. Kategori apa yang paling sering menjadi objek split bill dan berapa proporsinya?

## IMPORT LIBRARY

In [394]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## DATA GATHERING

In [395]:
data1 = pd.read_csv('data1.csv')
data1

,Date,Mode,Category,Subcategory,Note,Amount,Income/Expense,Currency
0,20/09/2018 12:04:08,Cash,Transportation,Train,2 Place 5 to Place 0,30.0,Expense,INR
1,20/09/2018 12:03:15,Cash,Food,snacks,Idli medu Vada mix 2 plates,60.0,Expense,INR
2,19/09/2018,Saving Bank account 1,subscription,Netflix,1 month subscription,199.0,Expense,INR
3,17/09/2018 23:41:17,Saving Bank account 1,subscription,Mobile Service Provider,Data booster pack,19.0,Expense,INR
4,16/09/2018 17:15:08,Cash,Festivals,Ganesh Pujan,Ganesh idol,251.0,Expense,INR
...,...,...,...,...,...,...,...,...
2456,1/1/2015,Cash,Transportation,NaN,share jeep - Place T base to top,20.0,Expense,INR
2457,1/1/2015,Cash,Transportation,NaN,share auto - Place H to Place T base,20.0,Expense,INR
2458,1/1/2015,Cash,Transportation,NaN,bus - brc to Place H,30.0,Expense,INR
2459,1/1/2015,Cash,Food,NaN,tea,10.0,Expense,INR


dataset 1 terdiri atas 2461 baris dan 8 kolom

In [396]:
pip install openpyxl

Note: you may need to restart the kernel to use updated packages.


In [397]:
data2 = pd.read_excel('data2.xlsx')
data2

,Date,Description,Amount,Transaction Type,Category,Account Name,Month
0,2018-01-01,Amazon,11.11,debit,Shopping,Platinum Card,2018-01
1,2018-01-02,Mortgage Payment,1247.44,debit,Mortgage & Rent,Checking,2018-01
2,2018-01-02,Thai Restaurant,24.22,debit,Restaurants,Silver Card,2018-01
3,2018-01-03,Credit Card Payment,2298.09,credit,Credit Card Payment,Platinum Card,2018-01
4,2018-01-04,Netflix,11.76,debit,Movies & Dvds,Platinum Card,2018-01
...,...,...,...,...,...,...,...
801,2019-09-27,Biweekly Paycheck,2250.00,credit,Paycheck,Checking,2019-09
802,2019-09-28,BP,33.46,debit,Gas & Fuel,Platinum Card,2019-09
803,2019-09-28,Sheetz,4.27,debit,Gas & Fuel,Platinum Card,2019-09
804,2019-09-30,Starbucks,1.75,debit,Coffee Shops,Platinum Card,2019-09


dataset 2 terdiri atas 806 baris dan 7 kolom

## DATA ASSESING

### ASSESING DATA1

In [398]:
# melihat informasi tipe data pada data1
data1.info()

<class 'pandas.DataFrame'>
RangeIndex: 2461 entries, 0 to 2460
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            2461 non-null   str    
 1   Mode            2461 non-null   str    
 2   Category        2461 non-null   str    
 3   Subcategory     1826 non-null   str    
 4   Note            1940 non-null   str    
 5   Amount          2461 non-null   float64
 6   Income/Expense  2461 non-null   str    
 7   Currency        2461 non-null   str    
dtypes: float64(1), str(7)
memory usage: 311.5 KB


- kolom `Date` bertipe data str, seharunys bertipe data datetime
- terdapat missing values yang perlu diatasi pada kolom `Subcategory` dan `Note`

In [399]:
# mengecek jumlah nilai yang hilang pada data1
data1.isna().sum()

Date                0
Mode                0
Category            0
Subcategory       635
Note              521
Amount              0
Income/Expense      0
Currency            0
dtype: int64

terdapat lebih dari 600 missing values pada data1

In [400]:
# mengecek jumlah nilai yang duplikat pada data1
data1.duplicated().sum()

np.int64(9)

tidak terdapat data yang duplikat pada data1

In [401]:
data1.describe()

,Amount
count,2461.000000
mean,2751.145380
std,12519.615804
min,2.000000
25%,35.000000
50%,100.000000
75%,799.000000
max,250000.000000


pada data 1, transaksi tertinggi berada pada nilai 250000 (INR) dan transaksi terendah berada pada nilai 2 (INR) dengan rata-rata transaksi bernilai 2751,1 (INR)

### ASSESING DATA2

In [402]:
# melihat informasi tipe data pada data2
data2.info()

<class 'pandas.DataFrame'>
RangeIndex: 806 entries, 0 to 805
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              806 non-null    datetime64[us]
 1   Description       806 non-null    str           
 2   Amount            806 non-null    float64       
 3   Transaction Type  806 non-null    str           
 4   Category          806 non-null    str           
 5   Account Name      806 non-null    str           
 6   Month             806 non-null    str           
dtypes: datetime64[us](1), float64(1), str(5)
memory usage: 82.3 KB


tidak terdapat tipe data yang aneh pada data2

In [403]:
# melihat jumlah nilai yang hilang pada data2
data2.isna().sum()

Date                0
Description         0
Amount              0
Transaction Type    0
Category            0
Account Name        0
Month               0
dtype: int64

tidak terdapat missing values pada data2

In [404]:
# melihat jumlah nilai yang duplikat pada data2
data2.duplicated().sum()

np.int64(0)

tidak ditemukan adanya nilai yang duplikat pada data2 

In [405]:
data2.describe()

,Date,Amount
count,806,806.000000
mean,2018-11-12 00:00:00,273.391489
min,2018-01-01 00:00:00,1.750000
25%,2018-06-03 06:00:00,15.687500
50%,2018-11-11 00:00:00,37.480000
75%,2019-04-18 18:00:00,117.680000
max,2019-09-30 00:00:00,9200.000000
std,NaN,667.630374


pada data2, transaksi tertinggi berada pada nilai 9200 (USD) dan transaksi terendah berada pada nilai 1,75 (USD) dengan rata-rata transaksi bernilai 273,39 (USD).

## DATA CLEANING

In [406]:
data1.head()

,Date,Mode,Category,Subcategory,Note,Amount,Income/Expense,Currency
0,20/09/2018 12:04:08,Cash,Transportation,Train,2 Place 5 to Place 0,30.0,Expense,INR
1,20/09/2018 12:03:15,Cash,Food,snacks,Idli medu Vada mix 2 plates,60.0,Expense,INR
2,19/09/2018,Saving Bank account 1,subscription,Netflix,1 month subscription,199.0,Expense,INR
3,17/09/2018 23:41:17,Saving Bank account 1,subscription,Mobile Service Provider,Data booster pack,19.0,Expense,INR
4,16/09/2018 17:15:08,Cash,Festivals,Ganesh Pujan,Ganesh idol,251.0,Expense,INR


In [407]:
# mengganti format tanggal pada kolom `Date` menjadi format datetime dan menghapus timestamp
def preprocess_date(date_str):
    if pd.isna(date_str):
        return date_str
    
    date_str = str(date_str).strip()
    
    if " " in date_str:
        date_str = date_str.split(" ")[0]
        
    return date_str

data1['Date'] = data1['Date'].apply(preprocess_date)
data1['Date'] = pd.to_datetime(data1['Date'], dayfirst=True, format='mixed', errors='coerce')

# normalisasi untuk menghilangkan jam tetapi format tetap datetime
data1['Date'] = data1['Date'].dt.normalize()

data1['Date'].head()

0   2018-09-20
1   2018-09-20
2   2018-09-19
3   2018-09-17
4   2018-09-16
Name: Date, dtype: datetime64[us]

In [408]:
# melihat apakah ada missing values setelah tipe data pada kolom `Date` sudah diubah menjadi datetime
data1['Date'].isna().sum()

np.int64(0)

In [409]:
# menghapus kolom yang tidak diperlukan untuk analisis
data1.drop(columns=['Mode', 'Subcategory', 'Note', 'Currency'], inplace=True)
data1.head()


,Date,Category,Amount,Income/Expense
0,2018-09-20,Transportation,30.0,Expense
1,2018-09-20,Food,60.0,Expense
2,2018-09-19,subscription,199.0,Expense
3,2018-09-17,subscription,19.0,Expense
4,2018-09-16,Festivals,251.0,Expense


In [410]:
# mengganti nama kolom `Income/Expense` menjadi `Transaction Type`
data1.rename(columns={'Income/Expense': 'Transaction Type'}, inplace=True)
data1.head()

,Date,Category,Amount,Transaction Type
0,2018-09-20,Transportation,30.0,Expense
1,2018-09-20,Food,60.0,Expense
2,2018-09-19,subscription,199.0,Expense
3,2018-09-17,subscription,19.0,Expense
4,2018-09-16,Festivals,251.0,Expense


In [411]:
print(data1.columns.tolist())
print(data1.head(3))

['Date', 'Category', 'Amount', 'Transaction Type']
        Date        Category  Amount Transaction Type
0 2018-09-20  Transportation    30.0          Expense
1 2018-09-20            Food    60.0          Expense
2 2018-09-19    subscription   199.0          Expense


karena pada mata uang yang digunakan pada data1 masih dalam bentuk INR (Indian Rupee) sedangkan target pengguna website Finance Tracker adalah generasi muda Indonesia, maka nilai mata uangnya akan dikonversi ke dalam Rupiah mengikuti kurs pada tahun di dataset yaitu 2015-2018 menggunakan sumber kurs dari World Bank.

In [412]:
# === KURS RATA-RATA TAHUNAN HISTORIS ===
# Sumber: World Bank, Official Exchange Rate (PA.NUS.FCRF)
# Indikator: PA.NUS.FCRF | https://data.worldbank.org
exchange_rates = {
    # tahun: (IDR_per_USD, INR_per_USD)
    2015: (13389, 64.15),
    2016: (13308, 67.20),
    2017: (13381, 65.12),
    2018: (14237, 68.40),
    2019: (14015, 70.42),
}

# ekstrak tahun dari kolom Date
data1['Year'] = pd.to_datetime(data1['Date']).dt.year

# Verifikasi tahun unik — pastikan sesuai dengan kurs yang tersedia
print("Tahun unik di data1:", data1['Year'].unique())

# fungsi konversi INR → IDR
def convert_inr_to_idr(amount, year):
    """
    Mengkonversi INR ke IDR menggunakan cross rate historis tahunan.
    Formula: IDR = INR × (IDR_per_USD ÷ INR_per_USD)
    Sumber: World Bank, PA.NUS.FCRF
    """
    if year not in exchange_rates:
        raise ValueError(f"Kurs untuk tahun {year} tidak tersedia.")
    
    idr_per_usd, inr_per_usd = exchange_rates[year]
    idr_per_inr = idr_per_usd / inr_per_usd
    return amount * idr_per_inr

# terapkan konversi
data1['Amount_IDR'] = data1.apply(
    lambda row: convert_inr_to_idr(row['Amount'], row['Year']),
    axis=1
)

data1.head()

Tahun unik di data1: [2018 2017 2016 2015]


,Date,Category,Amount,Transaction Type,Year,Amount_IDR
0,2018-09-20,Transportation,30.0,Expense,2018,6244.298246
1,2018-09-20,Food,60.0,Expense,2018,12488.596491
2,2018-09-19,subscription,199.0,Expense,2018,41420.511696
3,2018-09-17,subscription,19.0,Expense,2018,3954.722222
4,2018-09-16,Festivals,251.0,Expense,2018,52243.961988


nilai transaksi pada data1 telah dikonversi ke dalam bentuk rupiah dah disimpan pada kolom baru bernama `Amount_IDR`

In [413]:
# menghapus kolom `Amount` dan `Year` pada data1 karena sudah tidak diperlukan lagi setelah konversi ke IDR
data1.drop(columns=['Amount', 'Year'], inplace=True)
data1.head()

,Date,Category,Transaction Type,Amount_IDR
0,2018-09-20,Transportation,Expense,6244.298246
1,2018-09-20,Food,Expense,12488.596491
2,2018-09-19,subscription,Expense,41420.511696
3,2018-09-17,subscription,Expense,3954.722222
4,2018-09-16,Festivals,Expense,52243.961988


In [414]:
# mengecek kembali apakah ada kolom yang memiliki missing values dan duplikat setelah proses pembersihan dan konversi
print("Missing values per kolom:", data1.isna().sum())
print("Jumlah duplikat:", data1.duplicated().sum())

Missing values per kolom: Date                0
Category            0
Transaction Type    0
Amount_IDR          0
dtype: int64
Jumlah duplikat: 85


In [415]:
# menghapus kolom yang duplikat
data1.drop_duplicates(inplace=True)
print("Jumlah Duplikat:", data1.duplicated().sum())
data1

Jumlah Duplikat: 0


,Date,Category,Transaction Type,Amount_IDR
0,2018-09-20,Transportation,Expense,6244.298246
1,2018-09-20,Food,Expense,12488.596491
2,2018-09-19,subscription,Expense,41420.511696
3,2018-09-17,subscription,Expense,3954.722222
4,2018-09-16,Festivals,Expense,52243.961988
...,...,...,...,...
2454,2015-01-01,Food,Expense,41742.790335
2455,2015-01-01,Transportation,Expense,29637.381138
2458,2015-01-01,Transportation,Expense,6261.418550
2459,2015-01-01,Food,Expense,2087.139517


setelah dibersihkan, data1 memiliki 2376 baris dan 5 kolom.

### CLEANING DATA2

In [416]:
data2.head()

,Date,Description,Amount,Transaction Type,Category,Account Name,Month
0,2018-01-01,Amazon,11.11,debit,Shopping,Platinum Card,2018-01
1,2018-01-02,Mortgage Payment,1247.44,debit,Mortgage & Rent,Checking,2018-01
2,2018-01-02,Thai Restaurant,24.22,debit,Restaurants,Silver Card,2018-01
3,2018-01-03,Credit Card Payment,2298.09,credit,Credit Card Payment,Platinum Card,2018-01
4,2018-01-04,Netflix,11.76,debit,Movies & Dvds,Platinum Card,2018-01


In [417]:
# menghapus kolom yang tidak diperlukan untuk proses analisis
data2.drop(columns = ['Description', 'Account Name', 'Month'], inplace=True)
data2.head()

,Date,Amount,Transaction Type,Category
0,2018-01-01,11.11,debit,Shopping
1,2018-01-02,1247.44,debit,Mortgage & Rent
2,2018-01-02,24.22,debit,Restaurants
3,2018-01-03,2298.09,credit,Credit Card Payment
4,2018-01-04,11.76,debit,Movies & Dvds


In [418]:
# mengganti isi kolom Transaction Type menjadi Expense/Income
data2['Transaction Type'] = data2['Transaction Type'].replace({'debit': 'Expense', 'credit': 'Income'})
data2.head()

,Date,Amount,Transaction Type,Category
0,2018-01-01,11.11,Expense,Shopping
1,2018-01-02,1247.44,Expense,Mortgage & Rent
2,2018-01-02,24.22,Expense,Restaurants
3,2018-01-03,2298.09,Income,Credit Card Payment
4,2018-01-04,11.76,Expense,Movies & Dvds


In [419]:
# melakukan konversi nilai mata uang pada data2 dari USD ke IDR menggunakan kurs rata-rata tahunan historis
exchange_rates = {
    # tahun: (IDR_per_USD, INR_per_USD)
    2015: (13389, 64.15),
    2016: (13308, 67.20),
    2017: (13381, 65.12),
    2018: (14237, 68.40),
    2019: (14015, 70.42),
}

# ekstrak tahun dari kolom Date
data2['Year'] = pd.to_datetime(data2['Date']).dt.year

# Verifikasi tahun unik — pastikan sesuai dengan kurs yang tersedia
print("Tahun unik di data2:", data2['Year'].unique())

def convert_usd_to_idr(amount, year):
    """
    Mengkonversi USD ke IDR menggunakan kurs historis tahunan.
    Formula: IDR = USD × IDR_per_USD
    Sumber: World Bank, PA.NUS.FCRF
    """
    if year not in exchange_rates:
        raise ValueError(f"Kurs untuk tahun {year} tidak tersedia.")
    
    idr_per_usd, _ = exchange_rates[year]
    return amount * idr_per_usd
        
# terapkan konversi
data2['Amount_IDR'] = data2.apply(
    lambda row: convert_usd_to_idr(row['Amount'], row['Year']),
    axis=1
)

data2.head()

Tahun unik di data2: [2018 2019]


,Date,Amount,Transaction Type,Category,Year,Amount_IDR
0,2018-01-01,11.11,Expense,Shopping,2018,158173.07
1,2018-01-02,1247.44,Expense,Mortgage & Rent,2018,17759803.28
2,2018-01-02,24.22,Expense,Restaurants,2018,344820.14
3,2018-01-03,2298.09,Income,Credit Card Payment,2018,32717907.33
4,2018-01-04,11.76,Expense,Movies & Dvds,2018,167427.12


In [420]:
# menghapus kolom `Amount` dan `Year` yang sudah tidak diperlukan
data2.drop(columns=['Amount', 'Year'], inplace=True)
data2.head()

,Date,Transaction Type,Category,Amount_IDR
0,2018-01-01,Expense,Shopping,158173.07
1,2018-01-02,Expense,Mortgage & Rent,17759803.28
2,2018-01-02,Expense,Restaurants,344820.14
3,2018-01-03,Income,Credit Card Payment,32717907.33
4,2018-01-04,Expense,Movies & Dvds,167427.12


In [421]:
# mengecek kembali missing values dan duplikat setelah proses pembersihan dan konversi
print("Missing values per kolom:", data2.isna().sum())
print("Jumlah duplikat:", data2.duplicated().sum())

Missing values per kolom: Date                0
Transaction Type    0
Category            0
Amount_IDR          0
dtype: int64
Jumlah duplikat: 0


setelah dilakukan pembersihan dan konversi, tidak terdapat missing values ataupun duplikasi pada data2. karena kedua data sudah dalam keadaan bersih, maka data siap untuk dilakukan merge.

In [422]:
# melihat struktur data sebelum dilakukan merge
print("Kolom data1:", data1.columns.tolist())
print("Kolom data2", data2.columns.tolist())
print("Jumlah baris data1:", len(data1))
print("Jumlah baris data2:", len(data2))

Kolom data1: ['Date', 'Category', 'Transaction Type', 'Amount_IDR']
Kolom data2 ['Date', 'Transaction Type', 'Category', 'Amount_IDR']
Jumlah baris data1: 2376
Jumlah baris data2: 806


In [423]:
# melakukan merge pada data1 dan data2
data_merged = pd.concat([data1, data2], ignore_index=True)
print("\nJumlah baris setelah merge:", len(data_merged))
data_merged.head()
data_merged.info()


Jumlah baris setelah merge: 3182
<class 'pandas.DataFrame'>
RangeIndex: 3182 entries, 0 to 3181
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              3182 non-null   datetime64[us]
 1   Category          3182 non-null   str           
 2   Transaction Type  3182 non-null   str           
 3   Amount_IDR        3182 non-null   float64       
dtypes: datetime64[us](1), float64(1), str(2)
memory usage: 150.8 KB


jumlah baris data setelah dilakukan merge adalah 3182 baris dan 4 kolom yaitu:
- `Date` bertipe data datetime berisikan tanggal
- `Category` bertipe data string berisikan kategori tiap transaksi
- `Transaction Type` bertipe data string berisikan jenis transaksi (pengeluaran atau pemasukan)
- `Amount_IDR` bertipe data float berisikan nilai transaksi yang telah dikonversikan ke dalam mata uang Rupiah

In [424]:
print("jumlah missing values:", data_merged.isna().sum())
print("jumlah duplikat:", data_merged.duplicated().sum())

jumlah missing values: Date                0
Category            0
Transaction Type    0
Amount_IDR          0
dtype: int64
jumlah duplikat: 0


data_merged tidak memiliki missing values dan duplikat setelah proses pembersihan, konversi, dan merge. Data sudah siap untuk analisis lebih lanjut.

In [425]:
# melihat informasi berbagai category pada kolom `Category`
data_merged['Category'].unique()

<ArrowStringArray>
[           'Transportation',                      'Food',
              'subscription',                 'Festivals',
                     'Other',          'Small Cap fund 2',
          'Small cap fund 1',                    'Family',
      'Equity Mutual Fund E',                   'Apparel',
     'Public Provident Fund',     'Saving Bank account 1',
                      'Gift',                    'Salary',
                 'Household', 'Dividend earned on Shares',
                  'Interest',            'Life Insurance',
                    'Beauty',                    'Health',
            'Money transfer',                      'maid',
                   'Culture',                'Tax refund',
                   'Tourism',              'Share Market',
          'Self-development',       'Amazon pay cashback',
                 'Education',                     'scrap',
                'Petty cash',                 'Documents',
               'Gpay Reward',        

In [426]:
# melakukan mapping pada kolom `Category` untuk menyatukan kategori yang serupa antara data1 dan data2  
category_mapping = {
    # Income & Rewards
    'Salary': 'Income & Rewards', 'Dividend earned on Shares': 'Income & Rewards', 
    'Interest': 'Income & Rewards', 'Tax refund': 'Income & Rewards', 
    'Amazon pay cashback': 'Income & Rewards', 'Gpay Reward': 'Income & Rewards', 
    'Maturity amount': 'Income & Rewards', 'Bonus': 'Income & Rewards', 
    'Paycheck': 'Income & Rewards',
    
    # Food & Beverage
    'Food': 'Food & Beverage', 'Restaurants': 'Food & Beverage', 
    'Groceries': 'Food & Beverage', 'Fast Food': 'Food & Beverage', 
    'Coffee Shops': 'Food & Beverage', 'Food & Dining': 'Food & Beverage', 
    'Alcohol & Bars': 'Food & Beverage',
    
    # Housing & Utilities
    'Household': 'Housing & Utilities', 'maid': 'Housing & Utilities', 
    'Rent': 'Housing & Utilities', 'Cook': 'Housing & Utilities', 
    'garbage disposal': 'Housing & Utilities', 'water (jar /tanker)': 'Housing & Utilities', 
    'Mortgage & Rent': 'Housing & Utilities', 'Home Improvement': 'Housing & Utilities', 
    'Utilities': 'Housing & Utilities', 'Mobile Phone': 'Housing & Utilities', 
    'Internet': 'Housing & Utilities',
    
    # Transportation
    'Transportation': 'Transportation', 'Gas & Fuel': 'Transportation', 
    'Auto Insurance': 'Transportation',
    
    # Lifestyle & Entertainment
    'subscription': 'Lifestyle & Entertainment', 'Festivals': 'Lifestyle & Entertainment', 
    'Apparel': 'Lifestyle & Entertainment', 'Gift': 'Lifestyle & Entertainment', 
    'Culture': 'Lifestyle & Entertainment', 'Tourism': 'Lifestyle & Entertainment', 
    'Self-development': 'Lifestyle & Entertainment', 'Education': 'Lifestyle & Entertainment', 
    'Social Life': 'Lifestyle & Entertainment', 'Shopping': 'Lifestyle & Entertainment', 
    'Movies & Dvds': 'Lifestyle & Entertainment', 'Music': 'Lifestyle & Entertainment', 
    'Entertainment': 'Lifestyle & Entertainment', 'Television': 'Lifestyle & Entertainment', 
    'Electronics & Software': 'Lifestyle & Entertainment',
    
    # Health & Personal Care
    'Beauty': 'Health & Personal Care', 'Health': 'Health & Personal Care', 
    'Grooming': 'Health & Personal Care', 'Haircut': 'Health & Personal Care',
    
    # Investments & Savings
    'Small Cap fund 2': 'Investments & Savings', 'Small cap fund 1': 'Investments & Savings', 
    'Equity Mutual Fund E': 'Investments & Savings', 'Public Provident Fund': 'Investments & Savings', 
    'Life Insurance': 'Investments & Savings', 'Share Market': 'Investments & Savings', 
    'Equity Mutual Fund A': 'Investments & Savings', 'Fixed Deposit': 'Investments & Savings', 
    'Equity Mutual Fund C': 'Investments & Savings', 'Equity Mutual Fund F': 'Investments & Savings', 
    'Recurring Deposit': 'Investments & Savings', 'Equity Mutual Fund D': 'Investments & Savings', 
    'Equity Mutual Fund B': 'Investments & Savings', 'Investment': 'Investments & Savings',
    
    # Transfers & Debt Payment
    'Saving Bank account 1': 'Transfers & Debt Payment', 'Money transfer': 'Transfers & Debt Payment', 
    'Saving Bank account 2': 'Transfers & Debt Payment', 'Credit Card Payment': 'Transfers & Debt Payment',
    
    # Others
    'Other': 'Others', 'Family': 'Others', 'scrap': 'Others', 
    'Petty cash': 'Others', 'Documents': 'Others'
}
data_merged['Macro_Category'] = data_merged['Category'].map(category_mapping)

print("Distribusi Kategori Makro:")
print(data_merged['Macro_Category'].value_counts(dropna=False))

# mengecek apakah ada kategori yang terlewat (bocor menjadi NaN)
missing_mapping = data_merged[data_merged['Macro_Category'].isnull()]['Category'].unique()
if len(missing_mapping) > 0:
    print("\n[PERINGATAN] Ada kategori lama yang belum terpetakan dan menjadi NaN:")
    print(missing_mapping)
else:
    print("\n[SUKSES] Semua kategori berhasil dipetakan ke Kategori Makro!")

Distribusi Kategori Makro:
Macro_Category
Food & Beverage              1146
Lifestyle & Entertainment     384
Housing & Utilities           367
Transportation                346
Investments & Savings         243
Others                        207
Transfers & Debt Payment      206
Income & Rewards              150
Health & Personal Care        133
Name: count, dtype: int64

[SUKSES] Semua kategori berhasil dipetakan ke Kategori Makro!


mapping bertujuan untuk mengelompokkan kategori serupa yang ada pada data1 dan data2. pengelompokkan dibagi menjadi 9 kategori yang bisa dilihat pada output diatas.

In [427]:
# melihat informasi data pada data_merged setelah dilakukan mapping
print("Informasi Data Setelah Mapping:", data_merged.info())
print("Jumlah missing values:", data_merged.isna().sum())
print("Jumlah duplikat:", data_merged.duplicated().sum())

<class 'pandas.DataFrame'>
RangeIndex: 3182 entries, 0 to 3181
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              3182 non-null   datetime64[us]
 1   Category          3182 non-null   str           
 2   Transaction Type  3182 non-null   str           
 3   Amount_IDR        3182 non-null   float64       
 4   Macro_Category    3182 non-null   str           
dtypes: datetime64[us](1), float64(1), str(3)
memory usage: 230.0 KB
Informasi Data Setelah Mapping: None
Jumlah missing values: Date                0
Category            0
Transaction Type    0
Amount_IDR          0
Macro_Category      0
dtype: int64
Jumlah duplikat: 0


In [428]:
# menyimpan data yang sudah diproses ke dalam file CSV baru
data_merged.to_csv('data_merged.csv', index=False)

In [429]:
data_merged

,Date,Category,Transaction Type,Amount_IDR,Macro_Category
0,2018-09-20,Transportation,Expense,6.244298e+03,Transportation
1,2018-09-20,Food,Expense,1.248860e+04,Food & Beverage
2,2018-09-19,subscription,Expense,4.142051e+04,Lifestyle & Entertainment
3,2018-09-17,subscription,Expense,3.954722e+03,Lifestyle & Entertainment
4,2018-09-16,Festivals,Expense,5.224396e+04,Lifestyle & Entertainment
...,...,...,...,...,...
3177,2019-09-27,Paycheck,Income,3.153375e+07,Income & Rewards
3178,2019-09-28,Gas & Fuel,Expense,4.689419e+05,Transportation
3179,2019-09-28,Gas & Fuel,Expense,5.984405e+04,Transportation
3180,2019-09-30,Coffee Shops,Expense,2.452625e+04,Food & Beverage
